runtime_services/src/runtime_services.rs

```diff
@@ -0,0 +1,864 @@
+
+    /// Gets the name and namespace of the UEFI variable after the one provided.
+    ///
+    /// Returns a tuple of (name, namespace)
+    ///
+    /// Note: Unlike get_variable, a non-null terminated name will return INVALID_PARAMETER per UEFI spec
+    ///
+    /// UEFI Spec Documentation:
+    /// <a href="https://uefi.org/specs/UEFI/2.10/08_Services_Runtime_Services.html#getnextvariablename" target="_blank">
+    ///   8.2.2. EFI_RUNTIME_SERVICES.GetNextVariableName()
+    /// </a>
+    fn get_next_variable_name(
+        &self,
+        prev_name: &[u16],
+        prev_namespace: &efi::Guid,
+    ) -> Result<(Vec<u16>, efi::Guid), efi::Status> {
+        if prev_name.len() == 0 {
+            panic!("Zero-length name passed into get_next_variable_name.");
+        }
+
+        let mut next_name = Vec::<u16>::new();
+        let mut next_namespace: efi::Guid = efi::Guid::from_bytes(&[0x0; 16]);
+
+        unsafe {
+            self.get_next_variable_name_unchecked(&prev_name, &prev_namespace, &mut next_name, &mut next_namespace)?;
+        };
+
+        Ok((next_name, next_namespace))
+    }
+
+    /// Queries variable information for given UEFI variable attributes.
+    ///
+    /// UEFI Spec Documentation:
+    /// <a href="https://uefi.org/specs/UEFI/2.10/08_Services_Runtime_Services.html#queryvariableinfo" target="_blank">
+    ///   8.2.4. EFI_RUNTIME_SERVICES.QueryVariableInfo()
+    /// </a>
+    fn query_variable_info(&self, attributes: u32) -> Result<VariableInfo, efi::Status>;
+
+    /// Set's a UEFI variable
+    ///
+    /// # Safety
+    ///
+    /// Ensure name is null-terminated
+    unsafe fn set_variable_unchecked(
+        &self,
+        name: &mut [u16],
+        namespace: &efi::Guid,
+        attributes: u32,
+        data: &[u8],
+    ) -> Result<(), efi::Status>;
+
+    /// Gets a UEFI variable
+    ///
+    /// # Safety
+    ///
+    /// Ensure name is null-terminated
+    unsafe fn get_variable_unchecked<'a>(
+        &self,
+        name: &mut [u16],
+        namespace: &efi::Guid,
+        data: Option<&'a mut [u8]>,
+    ) -> GetVariableStatus;
+
+    /// Gets the UEFI variable name after the one provided.
+    ///
+    /// Will populate next_name and next_namespace.
+    ///
+    /// # Safety
+    ///
+    /// Ensure name isn't empty. It can be an empty string,
+    /// but there must be some data.
+    ///
+    unsafe fn get_next_variable_name_unchecked(
+        &self,
+        prev_name: &[u16],
+        prev_namespace: &efi::Guid,
+        next_name: &mut Vec<u16>,
+        next_namespace: &mut efi::Guid,
+    ) -> Result<(), efi::Status>;
+}
+
+impl RuntimeServices for StandardRuntimeServices<'_> {
+    unsafe fn set_variable_unchecked(
+        &self,
+        name: &mut [u16],
+        namespace: &efi::Guid,
+        attributes: u32,
+        data: &[u8],
+    ) -> Result<(), efi::Status> {
+        let set_variable = self.efi_runtime_services().set_variable;
+        if set_variable as usize == 0 {

```

Comment:
Should we panic at all if variable services fail? I would think not, though in early development it will be helpful. Can we control this like a C assert where on debug builds it panics, but not release? Or are we avoiding that pattern in Rust @makubacki?

Reply:
We're just following a similar paradigm to what was already done in boot_services. The idea here is that nobody should be calling runtime_services functions without initialization, but @makubacki we can also change everything (including the calls in boot_services) to use debug_assert! if you feel that's more appropriate

Reply:
I'm of the opinion that this is more appropriate as a `debug_assert!`. Panic has its place, but I think we should reserve it for actual irrecoverable errors. In this case, I would ask, "Is the fact that `set_variable` was not initialized in the RT Services right now catastrophic?" Ultimately, I don't think that's for the wrapper to decide. Unexpected? Yes. `SetVariable()` should not be called until the variable write arch protocol is installed. Catastrophic? Maybe. If the entire boot hinges upon this `SetVariable()` call for some (probably bad design) reason.

Most custom panic handlers we have today output an error log message and then deadloop in both debug and release profile. We can of course change that behavior (and we might) but it will be written with the understanding that a truly impassable situation has occurred. I don't think this is one. `debug_assert!()` with a message that says something like "SetVariable() invoked without service initialization in RT Services table" seems better and return an error for release profile. The code could also check if the variable write arch protocol is installed if it wants to be more helpful, then:

- If installed: "Variable write arch protocol is installed but SetVariable() is not provided."
- If not installed: "Caller tried to incorrectly invoke SetVariable() before the variable write arch protocol is installed. Update caller code."

Reply:
I've switched out the panic!s with debug_assert!s